qdl.c

```diff
@@ -110,6 +112,306 @@ static enum qdl_storage_type decode_storage(const char *storage)
+ *
+ * The original blob (in @images[0]) is freed once it has been consumed.
+ *
+ * Returns: 0 if no archive was found, 1 if archive was decoded, -1 on error
+ */
+static int decode_programmer_archive(struct sahara_image *images)
+{
+	struct sahara_image *blob = &images[0];
+	struct cpio_newc_header *hdr;
+	size_t filesize;
+	size_t namesize;
+	char name[128];
+	char *save;
+	char *tok;
+	void *ptr = blob->ptr;
+	void *end = blob->ptr + blob->len;
+	long id;
+
+	if (blob->len < sizeof(*hdr) || memcmp(ptr, CPIO_MAGIC, 6))
+		return 0;
+
+	for (;;) {
+		if (ptr + sizeof(*hdr) > end) {
+			fprintf(stderr, "programmer archive is truncated\n");
+			return -1;
+		}
+		hdr = ptr;
+
+		if (memcmp(hdr->c_magic, "070701", 6)) {
+			fprintf(stderr, "expected cpio header in programmer archive\n");
+			return -1;
+		}
+
+		filesize = parse_ascii_hex32(hdr->c_filesize);
+		namesize = parse_ascii_hex32(hdr->c_namesize);
+
+		ptr += sizeof(*hdr);
+		if (ptr + namesize > end || ptr + filesize + namesize > end) {
+			fprintf(stderr, "programmer archive is truncated\n");
+			return -1;
+		}
+
+		if (namesize > sizeof(name)) {
+			fprintf(stderr, "unexpected filename length in progammer archive\n");
+			return -1;
+		}
+		memcpy(name, ptr, namesize);
+
+		if (!memcmp(name, "TRAILER!!!", 11))
+			break;
+
+		tok = strtok_r(name, ":", &save);
+		id = strtoul(tok, NULL, 0);
+		if (id == 0 || id >= MAPPING_SZ) {
+			fprintf(stderr, "invalid image id \"%s\" in programmer archive\n", tok);
+			return -11;
+		}
+
+		ptr += namesize;
+		ptr = (void *)(((uintptr_t)ptr + 3) & ~3);
+
+		tok = strtok_r(NULL, ":", &save);
+		if (tok)
+			images[id].name = strdup(tok);
+		images[id].len = filesize;
+		images[id].ptr = malloc(filesize);
+		memcpy(images[id].ptr, ptr, filesize);
+
+		ptr += filesize;
+		ptr = (void *)(((uintptr_t)ptr + 3) & ~3);
+	}
+
+	free(blob->ptr);
+	blob->ptr = NULL;
+	blob->len = 0;
+
+	return 1;
+}
+
+/**
+ * decode_sahara_config() - Attempt to decode a Sahara config XML document
+ * @images: List of Sahara images, with @images[0] populated
+ *
+ * The single blob provided in @images[0] might be a XML blob containing
+ * a sahara_config document with definitions of the various Sahara images that
+ * will be loaded. Attempt to parse this and if possible load each referenced
+ * Sahara image into the @images array.
+ *
+ * The original blob (in @images[0]) is freed once it has been consumed.
+ *
+ * Returns: 0 if no archive was found, 1 if archive was decoded, -1 on error
+ */
+static int decode_sahara_config(struct sahara_image *images)
+{
+	struct sahara_image *blob = &images[0];
+	char image_path_full[PATH_MAX];
+	const char *image_path;
+	unsigned int image_id;
+	size_t image_path_len;
+	xmlNode *images_node;
+	xmlNode *image_node;
+	char *blob_name_buf;
+	size_t base_path_len;
+	char *base_path;
+	xmlNode *root;
+	xmlDoc *doc;
+	int errors = 0;
+	int ret;
+
+	if (blob->len < 5 || memcmp(blob->ptr, "<?xml", 5))
```

Comment:
I believe this is redundant, as `xmlReadMemory()` has all these checks within its implementation.

Reply:
I don't see this promise in the API specification, am I just missing it?